R/table-object.R

```diff
@@ -950,3 +950,33 @@ print.stobject <- function(x, ...) {
 #' @rdname print.stobject
 #' @export
 names.stobject <- function(x) names(x$data)
+
+
+#' Clone an stobject
+#'
+#' @param x an stobject object.
+#'
+#' @return
+#' A deep copy of `x`.
+#'
+#' @examples
+#' x <- st_new(stdata())
+#' y <- st_clone(x)
+#'
+#' y$data$STUDY <- NULL
+#' x$data
+#' y$data
+#'
+#' # Get back to a regular environment
+#' class(x) <- "environment"
+#' class(y) <- "environment"
+#'
+#' x
+#' y
+#'
+#' @export
+st_clone <- function(x) {
+  y <- list2env(as.list(x))
```

Comment:
For completeness, I think `as.list` should be called with `all.names = TRUE`.

Or perhaps use `rlang::env_clone` [*]?  That might take care of some remaining edge cases we're not thinking of and is faster (though I don't think it really matters in this context).

``` r
st_clone_1 <- function(x) {
  y <- list2env(as.list(x))
  attributes(y) <- attributes(x)
  y
}

st_clone_2 <- function(x) {
  y <- rlang::env_clone(x)
  attributes(y) <- attributes(x)
  y
}

a <- new.env(parent = emptyenv())
a$foo <- 1:10

rbenchmark::benchmark(
  st_clone_1(a),
  st_clone_2(a)
)
#>            test replications elapsed relative user.self sys.self user.child
#> 1 st_clone_1(a)          100   0.019     4.75     0.018    0.001          0
#> 2 st_clone_2(a)          100   0.004     1.00     0.004    0.000          0
#>   sys.child
#> 1         0
#> 2         0
```

[*] Note: `env_clone` doesn't do a recursive copy either.

Reply:
Thanks. I picked up `env_clone`